### ISSUE TITLE
Change library target from ES6 to ES5

### ISSUE THREAD

**Issue by vogloblinsky** [dsherret/ts-simple-ast#110]
Could you target ES5 for typescript output ?

I will try using it in my project than runs on Node.js 4. https://github.com/compodoc/compodoc

Thanks

**Comment by dsherret** [comment]
@vogloblinsky does specifying that as a compiler option not work? https://dsherret.github.io/ts-simple-ast/setup/index

**Comment by dsherret** [comment]
Oh, do you mean making this library itself target ES5? Right now it's targeting ES6 because it uses [Map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map) in [KeyValueCahce.ts](https://github.com/dsherret/ts-simple-ast/blob/master/src/utils/KeyValueCache.ts). Perhaps I could target ES5 then implement some fallback mechanism for when Map doesn't exist... that wouldn't be so bad. Maybe create a key-value object then add a dummy key property to the values being stored in the object.

**Comment by vogloblinsky** [comment]
This job https://travis-ci.org/compodoc/compodoc/jobs/295336899 failed because of this file : ts-simple-ast/dist/utils.js:11:10

**Comment by dsherret** [comment]
Ah ok, it's keeping the ES6 language features in so that's why that's failing. Yeah, that's currently expected because the [system requirements](https://dsherret.github.io/ts-simple-ast/) are `ES6 (Node v6.10+)`. The main problem to solve is `Map` being used in *KeyValueCache.ts* though. I can work on moving this to target ES5 and then use Map if it's available.

**Comment by vogloblinsky** [comment]
Oh cool, it could be nice.
I continue to target Node.js 4 for a portion of users of my tool.

**Comment by dsherret** [comment]
This is implemented in 0.80.0. I cried a little writing this. There was more to change than I expected.

The performance is very slightly slower (tests take 33 seconds to run instead of 30) and the coverage decreased quite a bit because typescript generates a lot of code that's not run... I need to figure out how to get istanbul to only look at the typescript. Overall, not a big deal... this was definitely something that needed to be done :)

I managed to implement an O(1) map for ES5. I had to do a little magic though and it's somewhat unorthodox, but without this the library would be painfully slow: https://github.com/dsherret/ts-simple-ast/blob/259fbe30842f848c89c237cf0fe02fd3e4008ffe/src/utils/KeyValueCache.ts (ES5Map class)
